cache/eventsource.go

```diff
@@ -0,0 +1,74 @@
+package cache
+
+import (
+	"context"
+	"time"
+
+	rpctypes "go.etcd.io/etcd/api/v3/v3rpc/rpctypes"
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+// serveWatchEvents runs one upstream watch, feeds ring & demux, and restarts on lag.
+func serveWatchEvents(
+	ctx context.Context,
+	client *clientv3.Client,
+	prefix string,
+	eventSink func(*clientv3.Event),
+	dx *demux,
+	ring History,
+	setCompaction func(int64),
+	backoffStart time.Duration,
+	backoffMax time.Duration,
+) {
+	go func() {
+		backoff := backoffStart
+		for {
+			// resume from the oldest revision we still keep
+			oldest := ring.OldestRevision()
+			opts := []clientv3.OpOption{
+				clientv3.WithPrefix(),
+				clientv3.WithProgressNotify(),
+			}
+			if oldest > 0 {
+				opts = append(opts, clientv3.WithRev(oldest+1))
+			}
+
+			watchCh := client.Watch(ctx, prefix, opts...)
+
+			for wr := range watchCh {
+				// update compaction info on every frame
+				if cr := wr.CompactRevision; cr > 0 {
+					setCompaction(cr)
+					if cr >= ring.LatestRevision() {
```

Comment:
Please read documentation about CompactRevision field https://github.com/etcd-io/etcd/blob/5dfd6e05a4bd8e09debe31742dee4221ff391d0d/api/etcdserverpb/rpc.proto#L846-L854 
This field is only set when stream is compacted, so this block is not needed.